diff --git a/vlib/v/ast/types.v b/vlib/v/ast/types.v
index db646448a..35441c613 100644
--- a/vlib/v/ast/types.v
+++ b/vlib/v/ast/types.v
@@ -51,6 +51,7 @@ pub enum Language {
 	arm32 // 32-bit arm
 	rv64  // 64-bit risc-v
 	rv32  // 32-bit risc-v
+	s390x
 	wasm32
 }
 
@@ -75,6 +76,9 @@ pub fn pref_arch_to_table_language(pref_arch pref.Arch) Language {
 		.i386 {
 			.i386
 		}
+		.s390x {
+			.s390x
+		}
 		.js_node, .js_browser, .js_freestanding {
 			.js
 		}
diff --git a/vlib/v/pref/arch.c.v b/vlib/v/pref/arch.c.v
index e683586fb..8994e1275 100644
--- a/vlib/v/pref/arch.c.v
+++ b/vlib/v/pref/arch.c.v
@@ -8,6 +8,7 @@ pub enum Arch {
 	rv64  // 64-bit risc-v
 	rv32  // 32-bit risc-v
 	i386
+	s390x
 	js_node
 	js_browser
 	js_freestanding
@@ -48,6 +49,9 @@ pub fn arch_from_string(arch_str string) !Arch {
 		'x86_32', 'x32', 'i386', 'IA-32', 'ia-32', 'ia32' { // i386 recommended
 			return .i386
 		}
+		's390x' {
+			return .s390x
+		}
 		'js', 'js_node' {
 			return .js_node
 		}
